Smart Function Concepts (#4600)
SoftwareAG/c8y-docs: content/concepts/smart-function-concept-bundle/overview.md

---
weight: 10
title: Overview
layout: redirect
---

### What are smart functions? {#what-are-smart-functions}

> Smart functions are a common extension mechanism built into {{< product-c8y-iot >}}. They let you embed custom Javascript logic directly into platform components—without building or deploying a custom microservice.

Reviewer: Is "common extension mechanism" the right phrase here?  The extension approach differs across each of the cases.

Reviewer: I like consistent extension approach ( used in later paragraph, I modified mechanism to approach)

Author: Done

> Think of them as function-as-a-service for the Internet of Things: each smart function is a focused piece of logic that runs inside a platform component, triggered by the component at the right moment. You write the function, configure it in the UI, and the platform handles execution, security, and scaling.

Reviewer: You have already highlighted various aspects on how it differs from microservices. Might be also good to highlight there that scaling is auto-managed.

Author: Done

**AI-assisted development**
Smart functions are designed to work well with AI code generation. The function signatures are simple and well-typed, which makes them easy to describe to an AI tool and easy to review. Where components provide an in-platform editing experience, AI assistance is integrated directly into the editor, allowing you to generate, refine, and explain smart functions without leaving the platform.

**Secure by design**
Smart functions run in a sandboxed environment. They cannot access the host filesystem, open network connections, or read another tenant's data. Resource consumption is bounded by the platform. For details, see [Sandbox and limits](#sandbox-and-limits).

### How smart functions fit into {{< product-c8y-iot >}} {#how-smart-functions-fit}

- **Data Preparation** uses smart functions inside rules to transform inbound device messages before they are stored, enabling data normalization, enrichment, and filtering at ingestion time.
- **Streaming Analytics** uses smart functions as custom blocks inside models, allowing you to apply business logic to real-time data streams.

> - **thin-edge.io** uses smart functions inside flows to process messages at the edge, reducing bandwidth and enabling local processing before data reaches the cloud.

Reviewer: Might be good to link respective documentation of the component

Author: They're linked from the implementations page, this is just examples, so I think it's fine here

You create and manage rules, models, and flows through each component's UI or API. The smart functions inside them are written in Javascript and executed by the platform.

### Key advantages {#key-advantages}

**A consistent extension approach across the platform**
Smart functions follow the same concepts and patterns wherever they appear. Learn the model once, and apply it across Data Preparation, Streaming Analytics, thin-edge.io, and future components. This consistency also reduces the learning required to build end-to-end solutions: the same mental model, tooling, and development workflow applies whether you are writing logic for the cloud or the edge.

**Easy movement of logic between cloud and edge**
Because smart functions share the same language and API patterns across components, it is straightforward to move business logic between environments. A transformation written for Data Preparation in the cloud can be adapted for a thin-edge.io flow on the device, without switching languages or development tools.

**Common tooling for development and packaging**

> Smart functions use standard Javascript tooling. You can write them in any editor, use TypeScript for type safety, and bundle external libraries using any standard Javascript build tool. The same workflow and packaging approach applies regardless of which component you are targeting.

Reviewer 2: Query: Does 'Common tooling for development and packaging' contradict 'Integrated into platform components and UI' 
For example: 'You can write them in any editor' vs 'You write and manage them through the component's own UI, not through a separate development or deployment tool'
Is it common tooling for development if it is in a component's own UI?

Author: It means you can use common (ie industry standard) tooling, since they all support building them off platform. Not that the in-platform UI is common

**Automatically scalable and secure**
The platform handles execution of smart functions, automatically scaling them as needed and ensuring they run securely in a sandboxed environment. You can focus on writing your logic without worrying about infrastructure, security, or performance.

**AI-assisted development**
Smart functions are designed to work well with AI code generation. The function signatures are simple and well-typed, which makes them easy to describe to an AI tool and easy to review. Where components provide an in-platform editing experience, AI assistance is integrated directly into the editor, allowing you to generate, refine, and explain smart functions without leaving the platform.

**Secure by design**
Smart functions run in a sandboxed environment. They cannot access the host filesystem, open network connections, or read another tenant's data. Resource consumption is bounded by the platform. For details, see [Sandbox and limits](#sandbox-and-limits).

### How smart functions fit into {{< product-c8y-iot >}} {#how-smart-functions-fit}

> - **Data Preparation** uses smart functions inside rules to transform inbound device messages before they are stored, enabling data normalization, enrichment, and filtering at ingestion time.

Reviewer 2: Comment: Same comment from glossary update, 'before they are stored' could be misleading to users implying it is only stored, not part of the usual data stream.

Author: This I think is less problematic than the other one, it's true it's stored, but it doesn't imply (to me) that it's the *only* thing that happens

Reviewer 2: True...
but does this read better?

```suggestion
- **Data Preparation** uses smart functions inside rules to transform inbound device messages before sending the result into the platform, enabling data normalization, enrichment, and filtering at ingestion time.
```
